src/ingestion.rs

```diff
@@ -22,102 +22,18 @@ pub enum IngestionError {
-) -> Result<IngestionNeeded, DbErr> {
-    let last_ingested_ledger_sequence = last_ingested_ledger_sequence(quasar_database).await?;
-    let last_stellar_ledger_sequence = last_stellar_ledger_sequence(node_database).await?;
-
-    let ingestion_needed = if last_ingested_ledger_sequence != last_stellar_ledger_sequence {
-        IngestionNeeded::Yes {
-            last_ingested_ledger_sequence,
-        }
-    } else {
-        IngestionNeeded::No
-    };
-
-    Ok(ingestion_needed)
-}
-
-async fn last_ingested_ledger_sequence(
-    quasar_database: &DatabaseConnection,
-) -> Result<Option<i32>, DbErr> {
-    let last_ingested_ledger = Ledger::find()
-        .order_by_desc(ledger::Column::Sequence)
-        .one(quasar_database)
-        .await?;
-    let last_ingested_ledger_sequence = last_ingested_ledger.map(|ledger| ledger.sequence);
-    Ok(last_ingested_ledger_sequence)
-}
-
-async fn last_stellar_ledger_sequence(
-    node_database: &DatabaseConnection,
-) -> Result<Option<i32>, DbErr> {
-    let last_stellar_ledger = ledgerheaders::Entity::find()
-        .order_by_desc(ledgerheaders::Column::Ledgerseq)
-        .one(node_database)
-        .await?;
-    let last_stellar_ledger_sequence = last_stellar_ledger.and_then(|ledger| ledger.ledgerseq);
-    Ok(last_stellar_ledger_sequence)
-}
-
-fn setup_metrics(
-    metrics: Registry,
-) -> prometheus::core::GenericCounter<prometheus::core::AtomicU64> {
-    let ingested_ledgers_counter =
-        IntCounter::new("ingested_ledgers", "Number of ingested ledgers").unwrap();
-    metrics
-        .register(Box::new(ingested_ledgers_counter.clone()))
-        .expect("Failed to register counter");
-    ingested_ledgers_counter
-}
-
 pub async fn ingest(
     node_database: DatabaseConnection,
     quasar_database: DatabaseConnection,
-    ingestion: &Ingestion,
+    ingestion: Ingestion,
     metrics: Registry,
 ) {
-    let ingested_ledgers = setup_metrics(metrics);
-
-    loop {
-        sleep(ingestion).await;
-
-        let ingestion_needed = new_ledgers_available(&node_database, &quasar_database).await;
-
-        match ingestion_needed {
-            Ok(IngestionNeeded::Yes {
-                last_ingested_ledger_sequence,
-            }) => {
-                debug!("New ledgers available");
-                let ingestion_result = ingest_ledgers(
-                    &node_database,
-                    &quasar_database,
-                    last_ingested_ledger_sequence,
-                    &ingested_ledgers,
-                )
-                .await;
-
-                if let Err(error) = ingestion_result {
-                    error!("Error while ingesting ledgers: {:?}", error);
-                }
-            }
-            Ok(IngestionNeeded::No) => {}
-            Err(error) => {
-                error!("Error while checking for new ledgers: {}", error);
-            }
-        }
-    }
+    let ledgers = ledgers::ingest(&node_database, &quasar_database, &ingestion, &metrics);
```

Comment:
There is a potentially significant drawback that this architecture has - it'd be impossible to keep data fully consistent. The old, ledger-based approach had the benefit of it being possible to be wrapped in a single, giant transaction. With separate ingestion processes we're almost guaranteed to have inconsistent data throughout the ingestion process.

I was internally debating whether this is something we should worry about in the PoC, but if there were an easy way of achieving data consistency then we should go for it.

Reply:
You wanted to wrap the loop in a transaction?

Reply:
I wanted to wrap `ingest_ledger` in a transaction. This way if there's a new ledger all the new updates for all entities related to that new ledger get applied to the Quasar database in a single transaction. This way the data is always consistent.

Reply:
Cant we still do it with the current approach?

Reply:
The changes i did was just move around code.
I am ok if you have suggestions of what code changes make it impossible to do the transaction.

Reply:
If I understand correctly your code has a separate check for new ledgers for each new entity type. That check is done every 5 seconds for each entity. If I'm not mistaken the following scenario is possible:

1. Transaction ingestion check runs, but finds no new ledgers. No new ledgers are present so nothing is done.
2. 1 second later a new ledger is created.
3. 1 second later the account ingestion checks for a new ledger. It finds a new ledger so it imports accounts.
4. A users runs a GraphQL query to query for accounts. The users tries to join related transactions, but the transaction ingestion hasn't run yet as it'll run in a few seconds. The user receives inconsistent data - accounts without related transactions.

Reply:
If we have a single entry point to the entire ingestion process, ie. the check for a new ledger is executed in only one place then updates of all entities can be done all at once and encapsulated in a single transaction. So we should have (pseudocode):

```
loop {
  sleep();

  if new_ledger_present {
    start_transaction();
    
    ingest_ledgers();
    ingest_accounts();
    [... other ingestions ...]
    
    end_transaction();
  }
}
```

Reply:
I have made some updates, let me know what you think.